onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -287,88 +292,95 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
 
         for (int64_t i = 0; i < N; ++i) {
           score = {0, 0};
-          for (j = 0; j < static_cast<size_t>(n_trees_); ++j)
+          for (j = 0; j < static_cast<size_t>(n_trees_); ++j) {
             agg.ProcessTreeNodePrediction1(score, *ProcessTreeNodeLeave(roots_[j], x_data + i * stride));
+          }
+
           agg.FinalizeScores1(z_data + i * n_targets_or_classes_, score,
-                              label_data == NULL ? NULL : (label_data + i));
+                              label_data == nullptr ? nullptr : (label_data + i));
         }
       } else {
         concurrency::ThreadPool::TryBatchParallelFor(
             ttp,
-            static_cast<int32_t>(N),
-            [&](ptrdiff_t i) {
+            SafeInt<int32_t>(N),
+            [this, &agg, x_data, z_data, stride, label_data](ptrdiff_t i) {
               ScoreValue<OTYPE> score = {0, 0};
-              for (size_t j = 0; j < static_cast<size_t>(n_trees_); ++j)
+              for (size_t j = 0; j < static_cast<size_t>(n_trees_); ++j) {
                 agg.ProcessTreeNodePrediction1(score, *ProcessTreeNodeLeave(roots_[j], x_data + i * stride));
+              }
+
               agg.FinalizeScores1(z_data + i * n_targets_or_classes_, score,
-                                  label_data == NULL ? NULL : (label_data + i));
+                                  label_data == nullptr ? nullptr : (label_data + i));
             },
             0);
       }
     }
   } else {
     if (N == 1) {
       std::vector<ScoreValue<OTYPE>> scores(n_targets_or_classes_, {0, 0});
-
       if (n_trees_ <= parallel_tree_) {
-        for (int64_t j = 0; j < n_trees_; ++j)
+        for (int64_t j = 0; j < n_trees_; ++j) {
           agg.ProcessTreeNodePrediction(scores, *ProcessTreeNodeLeave(roots_[j], x_data));
-        agg.FinalizeScores(scores, z_data, -1, label_data);
+        }
       } else {
-        auto nth = n_trees_ * 2 / concurrency::ThreadPool::NumThreads(ttp);
-        if (n_trees_ % nth != 0)
-          ++nth;
-        concurrency::ThreadPool::TryBatchParallelFor(
+        // split the work into one block per thread so we can re-use the 'private_scores' vector as much as possible
+        // TODO: Refine the number of threads used
+        auto num_threads = std::min<int32_t>(concurrency::ThreadPool::NumThreads(ttp), SafeInt<int32_t>(n_trees_));
+        std::mutex merge_mutex;
+        concurrency::ThreadPool::TrySimpleParallelFor(
             ttp,
-            static_cast<int32_t>(nth),
-            [&](ptrdiff_t th) {
+            num_threads,
+            [this, &agg, &scores, &merge_mutex, num_threads, x_data](ptrdiff_t batch_num) {
               std::vector<ScoreValue<OTYPE>> private_scores(n_targets_or_classes_, {0, 0});
-              auto end = nth * (th + 1);
-              end = end < N ? end : N;
-              for (int64_t j = nth * th; j < end; ++j) {
+              int64_t j = batch_num * n_trees_ / num_threads;
```

Comment:
Would using PartitionWork like [here](https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/platform/threadpool.h#L307-L313) be better so that other ops that need to do this can follow the same partitioning logic that comes from the threadpool interface?

Reply:
Not sure. Both return the same values. One is faster. The other is easier to understand. It's also a little painful to have to flip between ptrdiff_t and other types depending on where PartitionWork is called from. Maybe we could have a templatized version that does the 2 line calculation with doco to explain it.

Python to prove the 2 line method works (which is what we originally had in the threadpool implementation):

```
for total in range(3, 1000, 17):
    for num_batches in range(2, 1000, 3):
        if num_batches > total:
            continue

        processed = 0
        summary = {}
        for i in range(num_batches):
            start = int(i * total/num_batches)
            end = int((i + 1) * total / num_batches)
            items = end - start
            processed += items
            if items in summary:
                summary[items] += 1
            else:
                summary[items] = 1

        # if not evenly divisible there should be only two batch sizes
        assert(len(summary) <= 2)
        if len(summary) == 2:
            # if two batch sizes they should only differ by 1
            sizes = sorted(summary.keys())
            assert(sizes[1] - sizes[0] == 1)

        assert(processed == total)

```

Reply:
Sure, that's fine.